### PR TITLE
🎧 Add Multi-Band MelGAN with ParallelWaveGAN discriminator (no master commits)

### PR DESCRIPTION
**Pull request by ZDisket**
In my experiments, combining mb-melgan generator (loading weights from pretrained model) with pwgan discriminator (from scratch) yields good results in very few steps (from 500 to 5000), exponentially increasing speed and quality of finetuning (these results are much better and resistant to differences than just finetuning regular mb-melgan or pwgan).
This will allow for very fast training vocoders.

### CODE REVIEW

File: examples/multiband_pwgan/README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Multi-band MelGAN: Faster Waveform Generation for High-Quality Text-to-Speech (With ParallelWaveGAN discriminator)
+Based on the script [`train_multiband_pwgan.py`](https://github.com/dathudeptrai/TensorflowTTS/tree/master/examples/multiband_pwgan/train_multiband_pwgan.py).
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:2]: `dathudeptrai` -> `tensorspeech`

File: examples/multiband_pwgan/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+## Training Multi-band MelGAN from scratch with LJSpeech dataset.
+This example code show you how to train MelGAN from scratch with Tensorflow 2 based on custom training loop and tf.function. The data used for this example is LJSpeech, you can download the dataset at  [link](https://keithito.com/LJ-Speech-Dataset/).
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:5]: This example code show you how to train MB-MelGAN with PWG discriminator ...

File: examples/multiband_pwgan/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+### Step 1: Create Tensorflow based Dataloader (tf.dataset)
+Please see detail at [examples/melgan/](https://github.com/dathudeptrai/TensorflowTTS/tree/master/examples/melgan#step-1-create-tensorflow-based-dataloader-tfdataset)
+
+### Step 2: Training from scratch
+After you re-define your dataloader, pls modify an input arguments, train_dataset and valid_dataset from [`train_multiband_pwgan.py`](https://github.com/dathudeptrai/TensorflowTTS/tree/master/examples/multiband_pwgan/train_multiband_pwgan.py). Here is an example command line to training melgan-stft from scratch:
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:11]: `dathudeptrai` -> `tensorspeech`

File: examples/multiband_pwgan/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+First, you need training generator with only stft loss: 
+
+```bash
+CUDA_VISIBLE_DEVICES=0 python examples/multiband_pwgan/train_multiband_pwgan.py \
+  --train-dir ./dump/train/ \
+  --dev-dir ./dump/valid/ \
+  --outdir ./examples/multiband_pwgan/exp/train.multiband_melgan.v1/ \
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:19]: maybe need rename :)). `train.multiband_melgan` -> `train.multiband_melgan_pwg_dis.v1` for example

File: examples/multiband_pwgan/README.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  --resume ""
+```
+
+Then resume and start training generator + discriminator:
+
+```bash
+CUDA_VISIBLE_DEVICES=0 python examples/multiband_pwgan/train_multiband_pwgan.py \
+  --train-dir ./dump/train/ \
+  --dev-dir ./dump/valid/ \
+  --outdir ./examples/multiband_pwgan/exp/train.multiband_melgan.v1/ \
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:32]: rename also

File: examples/multiband_pwgan/README.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  --config ./examples/multiband_pwgan/conf/multiband_pwgan.v1.yaml \
+  --use-norm 1 \
+  --resume ./examples/multiband_pwgan/exp/train.multiband_melgan.v1/checkpoints/ckpt-200000
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:35]: rename for consistent

File: examples/multiband_pwgan/README.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+IF you want to use MultiGPU to training you can replace `CUDA_VISIBLE_DEVICES=0` by `CUDA_VISIBLE_DEVICES=0,1,2,3` for example. You also need to tune the `batch_size` for each GPU (in config file) by yourself to maximize the performance. Note that MultiGPU now support for Training but not yet support for Decode. 
+
+In case you want to resume the training progress, please following below example command line:
+
+```bash
+--resume ./examples/multiband_pwgan/exp/train.multiband_melgan.v1/checkpoints/ckpt-100000
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:43]: rename for consistent

File: examples/multiband_pwgan/README.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+- If Your Dataset is 16K, upsample_scales = [2, 4, 8] worked.
+- If Your Dataset is > 16K (22K, 24K, ...), upsample_scales = [2, 4, 8] didn't worked, used [8, 4, 2] instead.
+
+### Step 3: Decode audio from folder mel-spectrogram
+To running inference on folder mel-spectrogram (eg valid folder), run below command line:
+
+```bash
+CUDA_VISIBLE_DEVICES=0 python examples/multiband_pwgan/decode_mb_melgan.py \
+  --rootdir ./dump/valid/ \
+  --outdir ./prediction/multiband_melgan.v1/ \
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:57]: rename :v

File: examples/multiband_pwgan/README.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  --checkpoint ./examples/multiband_pwgan/exp/train.multiband_melgan.v1/checkpoints/generator-940000.h5 \
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:58]: rename :v

File: examples/multiband_pwgan/README.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+## Finetune Multi-Band MelGAN + PWGAN Disc with ljspeech pretrained on other languages
+Download generator weights of Multi-Band MelGAN model, pass to `--pretrained` argument
+
+```bash
+CUDA_VISIBLE_DEVICES=0 python examples/multiband_pwgan/train_multiband_pwgan.py \
+  --train-dir ./dump/train/ \
+  --dev-dir ./dump/valid/ \
+  --outdir ./examples/multiband_pwgan/exp/train.multiband_melgan.v1ft/ \
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:71]: rename :v

File: examples/multiband_pwgan/README.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+  --config ./examples/multiband_pwgan/conf/multiband_pwgan.v1.yaml \
+  --use-norm 1 \
+  --generator_mixed_precision 1 \
+  --pretrained "ptgen.h5"
+```
+## Learning Curves
+Here is a learning curves of melgan based on this config [`multiband_melgan.v1.yaml`](https://github.com/dathudeptrai/TensorflowTTS/tree/master/examples/multiband_pwgan/conf/multiband_pwgan.v1.yaml)
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:78]: dathudeptrai -> tensorspeech.

File: examples/multiband_pwgan/README.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+<img src="fig/eval.png" height="300" width="850">
+
+<img src="fig/train.png" height="300" width="850">
+
+## Pretrained Models and Audio samples
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/README.md:84]: Remove this section since we not yet provide pretrained.

File: examples/multiband_pwgan/conf/multiband_pwgan.v1ft.yaml [hunk 13]
```diff
@@ -0,0 +1,99 @@
+
```
Comment (dathudeptrai) [inline examples/multiband_pwgan/conf/multiband_pwgan.v1ft.yaml:1]: write a note that this config is tune on ur dataset so user need to adjust by theirself in their case. The differences are `discriminator_train_start_steps` = 0 and learning rate = ... :D

**Comment by ZDisket** [comment]
Everything should be replaced and clarified.
